atlp-rwanda/ecommerce-app-predators-bn
# #1 ft(discount-coupon):Seller CRUD Operation functionalities

#### What does this PR do?
This pull request adds functionality to create, retrieve, update, and delete DiscountCoupons by ID.
#### Description of Task to be completed?

> Acceptance criteria:

> GIVEN a post request to create a new coupon code
WHEN the requesting user/token is a seller and the request body is validated
THEN the new coupon code should be saved to the database with the seller's account ID

> GIVEN a post request to get coupon codes
WHEN the requesting user/token is a seller
THEN the backend should retrieve all coupon codes associated with the seller's account ID and return them to the frontend for display in the "Coupons" section of the seller's dashboard
#### How should this be manually tested?

> To manually test the CRUD operations for coupons, you can follow the following steps:

`git clone branch ft-seller-discount-coupons-1`
`run npm install`
`run npm run dev`

> Use any Testing Tool
**To Create a coupon you have to be a seller make sure you have right credentials**

> http://localhost:3000/api/discount-coupons/createCoupon
**To get Created Coupons **

> http://localhost:3000/api/discount-coupons/getCoupons
**To Update a Coupon by Id**

> http://localhost:3000/api/discount-coupons/updateCoupon/:id
**To delete a Coupon by Id**

> http://localhost:3000/api/discount-coupons/deleteCoupons/:id

Create a new coupon by sending a POST request to the API endpoint with the required fields (code, discountPercentage, expiresAt, and productId) in the request body. You should receive a response with the created coupon details in JSEND format and a status code of 200 or 201.

Retrieve a coupon by its ID by sending a GET request to the API endpoint with the coupon ID as a URL parameter. You should receive a response with the coupon details in JSEND format and a status code of 200.

Update a coupon by sending a PUT or PATCH request to the API endpoint with the coupon ID as a URL parameter and the fields to be updated in the request body. You should receive a response with the updated coupon details in JSEND format and a status code of 200.

Delete a coupon by sending a DELETE request to the API endpoint with the coupon ID as a URL parameter. You should receive a response with a success message in JSEND format and a status code of 200.

You can use tools like Postman, cURL or HTTPie to manually test these API endpoints.

#### What are the relevant pivotal tracker/Trello stories?

> https://trello.com/c/sqzTHFBN/1-seller-should-be-able-to-create-discount-coupons
#### Screenshots (if appropriate)
<img width="1601" alt="Screenshot 2023-04-30 at 13 37 52" src="https://user-images.githubusercontent.com/122970729/235350983-5e8cc39a-3a32-49d2-b858-405b0489b862.png">
